src/apps/bluebit/aurora/bluebit.vault.token-fetcher.ts

```diff
@@ -0,0 +1,68 @@
+import { Inject } from '@nestjs/common';
+import { uniq } from 'lodash';
+
+import { IAppToolkit, APP_TOOLKIT } from '~app-toolkit/app-toolkit.interface';
+import { Register } from '~app-toolkit/decorators';
+import { UniswapPair, UniswapV2ContractFactory, UniswapV2PoolTokenHelper } from '~apps/uniswap-v2';
+import { PositionFetcher } from '~position/position-fetcher.interface';
+import { AppTokenPosition } from '~position/position.interface';
+import { Network } from '~types/network.interface';
+
+import { BLUEBIT_DEFINITION } from '../bluebit.definition';
+import { Bluebit, BluebitContractFactory } from '../contracts';
+
+const appId = BLUEBIT_DEFINITION.id;
+const groupId = BLUEBIT_DEFINITION.groups.vault.id;
+const network = Network.AURORA_MAINNET;
+
+const zeroAddress = '0x0000000000000000000000000000000000000000';
+const statsAddress = '0x36C6FBA304009a036BaaE1a24a570B450Ae14a5C';
+
+@Register.TokenPositionFetcher({ appId, groupId, network })
+export class AuroraBluebitVaultTokenFetcher implements PositionFetcher<AppTokenPosition> {
+  constructor(
+    @Inject(UniswapV2PoolTokenHelper)
+    private readonly poolTokenHelper: UniswapV2PoolTokenHelper,
+    @Inject(UniswapV2ContractFactory) private readonly uniswapV2ContractFactory: UniswapV2ContractFactory,
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(BluebitContractFactory) private readonly bluebitContractFactory: BluebitContractFactory,
+  ) {}
+
+  async getVaults() {
+    const stats = this.bluebitContractFactory.stats({ address: statsAddress, network: network });
+    const vaults: any = await stats.vaults(zeroAddress);
+    let items: any = [];
+    for (let i = 0; i < vaults.length; i++) {
+      const pair = this.uniswapV2ContractFactory.uniswapPair({ address: vaults[i].token, network });
+      try {
+        await pair.token0();
+        items.push(vaults[i].token.toLowerCase());
+      } catch (error) {}
+    }
```

Comment:
Smart idea! You can leverage multicall here though so that these calls get batched together, here's an example:

```ts
const multicall = this.appToolkit.getMulticall(network);

const vaultAddresses = await Promise.all(vaults.map(async vault => {
  const pair = this.uniswapV2ContractFactory.uniswapPair({ address: vaults[i].token, network });
  // Check if the vault token is a valid Uniswap pair
  const maybePairToken0 = await multicall.request(pair).token().catch(() => null);
  return maybePairToken0 ? vault.token.toLowerCase() : null;
}));

return _.compact(vaultAddresses); // compact to filter null values
```

Reply:
fixed